File: python/pig-latin/pig_latin.py
def index_noexcept(array, item):
	if item in array:
		return array.index(item)

	return -1

def vowel(character):
	return (-1 != index_noexcept(['a', 'e', 'i', 'o', 'u'], character))

def compose_helper(word, index):
	return word[index + 1:] + word[0:index + 1] + 'ay'

def translate_helper(word):
	h_pos = index_noexcept(word, 'h')
	q_pos = index_noexcept(word, 'q')

	if (-1 != h_pos):
		if (('r' == word[h_pos + 1]) and ('t' == word[h_pos - 1])):
			return compose_helper(word, h_pos + 1)

		return compose_helper(word, h_pos)
	elif (('u' == word[q_pos + 1]) and (1 >= q_pos) or (-1 != q_pos)):
		return compose_helper(word, q_pos + 1)

	if vowel(word[0]):
		return word + 'ay'

	return compose_helper(word, 0)

def translate(english_word):
	# it's a phrase
	if (-1 != index_noexcept(english_word, ' ')):
		messed_phrase = ''
		words = english_word.split(' ')

		words_len = len(words)
		for i in range(words_len):
			messed_phrase += translate_helper(words[i]) + ' '

		return messed_phrase.strip()

	# it's a single word
	return translate_helper(english_word)

File: javascript/pig-latin/pig-latin.js
function PigLatin() {}

PigLatin.compose_helper = function(word, index) {
	return word.substring(index + 1) + word.substring(0, index + 1) + 'ay';
};

PigLatin.translate_helper = function(word) {
	var hPos = word.indexOf('h');
	var qPos = word.indexOf('q');
	if (-1 !== hPos) {
		if (('r' === word[hPos + 1]) && ('t' === word[hPos - 1])) {
			return this.compose_helper(word, hPos + 1);
		}
		
		return this.compose_helper(word, hPos);
	} else if (('u' === word[qPos + 1]) && (1 >= qPos) && (-1 !== qPos)) {
		return this.compose_helper(word, qPos + 1);
	}
		
	if (isVowel(word[0])) {
		return word + 'ay';
	}
	
	return this.compose_helper(word, 0);
};

PigLatin.translate = function(input) {
	// it's a phrase
	if (-1 !== input.indexOf(' ')) {
		var messedPhrase = '';
		var words = input.split(' ');
		
		var wordsLength = words.length;
		for (var i = 0; i < wordsLength; ++i) {
			messedPhrase += this.translate_helper(words[i]) + ' ';
		}
		
		return messedPhrase.trim();
	}
	
	// it's a single word
	return this.translate_helper(input);
};

function isVowel(character) {
	return (-1 !== ['a', 'e', 'i', 'o', 'u'].indexOf(character));
}

module.exports = PigLatin;